src/VisualStudio/Core/Def/Implementation/ProjectSystem/CPS/CPSProject_IWorkspaceProjectContext.cs

```diff
@@ -88,33 +92,47 @@ public new void RemoveMetadataReference(string referencePath)
             base.RemoveMetadataReference(referencePath);
         }
 
-        public void AddProjectReference(IProjectContext project, MetadataReferenceProperties properties)
+        public void AddProjectReference(IWorkspaceProjectContext project, MetadataReferenceProperties properties)
         {
+            var abstractProject = GetAbstractProject(project);
+
             // AbstractProject and ProjectTracker track project references using the project bin output path.
             // Setting the command line arguments should have already set the output file name and folder.
             // We fetch this output path to add the reference.
-            var referencedProject = this.ProjectTracker.GetProject(project.Id);
+            var referencedProject = this.ProjectTracker.GetProject(abstractProject.Id);
             var binPathOpt = referencedProject.TryGetBinOutputPath();
             if (!string.IsNullOrEmpty(binPathOpt))
             {
                 AddMetadataReferenceAndTryConvertingToProjectReferenceIfPossible(binPathOpt, properties);
             }
         }
 
-        public void RemoveProjectReference(IProjectContext project)
+        public void RemoveProjectReference(IWorkspaceProjectContext project)
+        {
+            var referencedProject = GetAbstractProject(project);
+
+            // AbstractProject and ProjectTracker track project references using the project bin output path.
+            // We fetch this output path to remove the reference.
+            var binPathOpt = referencedProject.TryGetBinOutputPath();
+            if (!string.IsNullOrEmpty(binPathOpt))
+            {
+                base.RemoveMetadataReference(binPathOpt);
+            }
+        }
+
+        private AbstractProject GetAbstractProject(IWorkspaceProjectContext project)
```

Comment:
Should this be static? Also, why not just have direct casts? This function implies it's doing something fancy by name that it's not. And I'll take an InvalidCastException over ArgumentException any day, as it's a bit more clear what the failure is.